packages/playwright-test/src/runner.ts

```diff
@@ -427,15 +452,11 @@ export class Runner {
       let sigintWatcher;
 
       let hasWorkerErrors = false;
-      let previousStageFailed = false;
-      // TODO: will be project setups followed by tests.
-      const orderedTestGroups = [testGroups];
-      for (let testGroups of orderedTestGroups) {
-        if (previousStageFailed)
-          testGroups = this._skipTestsNotMarkedAsRunAlways(testGroups);
-        if (!testGroups.length)
-          continue;
-        const dispatcher = new Dispatcher(this._loader, [...testGroups], this._reporter);
+      const failedProjects = new Set<string>();
+      for (let stageGroups of [projectSetupGroups, testGroups]) {
+        if (failedProjects.size)
+          stageGroups = this._skipTestsFromFailedProjects(stageGroups, failedProjects);
+        const dispatcher = new Dispatcher(this._loader, [...stageGroups], this._reporter);
```

Comment:
Let's consider extracting line 459-472 in a separate function and just call it twice for now.